src/ray/core_worker/task_manager.cc

```diff
@@ -2,10 +2,11 @@
 
 namespace ray {
 
-void TaskManager::AddPendingTask(const TaskSpecification &spec) {
+void TaskManager::AddPendingTask(const TaskSpecification &spec, int num_retries_allowed) {
   RAY_LOG(DEBUG) << "Adding pending task " << spec.TaskId();
   absl::MutexLock lock(&mu_);
-  RAY_CHECK(pending_tasks_.emplace(spec.TaskId(), spec.NumReturns()).second);
+  std::pair<TaskSpecification, int> entry = {spec, num_retries_allowed};
+  RAY_CHECK(pending_tasks_.emplace(spec.TaskId(), std::move(entry)).second);
```

Comment:
Could you please change `pending_tasks_`  to contain shared_ptr of `TaskSpec` instead, copying `TaskSpec` is expensive.

And also I think we only need to add the task spec into the map when  `num_retries_allowed` is non-zero, when it's set to 0 then we can pass a nullptr instead.

Reply:
TaskSpec already contains a shared_ptr to the protobuf, so it is meaningless to have another shared_ptr to TaskSpec on top of that.

We could do what you described eventually, but it's pretty awkward (you'd still have to save some info like number of returns so you can fail the task properly). I think this is something we can optimize separately after this PR is merged.

Reply:
Later on, perhaps we could just store a shared_ptr to the PushTaskRequest instead of a shared_ptr to the TaskSpec protobuf.

Reply:
OK, sounds good.

Could you please make sure we only cache the request/taskspec when max_retries is non-zero, for direct actor calls it doesn't have retries and thus it doesn't make sense to cache. If you prefer to do this in a later PR, could you please add a TODO to track?